### ISSUE TITLE
js_free leak

### ISSUE THREAD

**Issue by dzonerzy** [ccxvii/mujs#102]
Using master from github results in information leak due to js_free calling js_defaultalloc, this will somehow leak 2/1 bytes each time thus preventing libfuzzer from working without using -detect_leak=0

```
$ ./mujs test.js 
garbage collected: 0/1 envs, 2/5 funs, 4/362 objs, 0/0 strs

==============================================================
==1558==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 2 byte(s) in 2 object(s) allocated from:
    #0 0x7f7589ddaf40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x55678e058126 in js_defaultalloc /home/fuzzing/Desktop/mujs/jsstate.c:12
    #2 0x55678e04a1df in js_free /home/fuzzing/Desktop/mujs/jsrun.c:58
    #3 0x55678e029015 in jsG_freefunction /home/fuzzing/Desktop/mujs/jsgc.c:18
    #4 0x55678e02b2ad in js_freestate /home/fuzzing/Desktop/mujs/jsgc.c:242
    #5 0x55678e06d43e in main /home/fuzzing/Desktop/mujs/main.c:373
    #6 0x7f7589345b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Direct leak of 2 byte(s) in 2 object(s) allocated from:
    #0 0x7f7589ddaf40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x55678e058126 in js_defaultalloc /home/fuzzing/Desktop/mujs/jsstate.c:12
    #2 0x55678e04a1df in js_free /home/fuzzing/Desktop/mujs/jsrun.c:58
    #3 0x55678e028fd9 in jsG_freefunction /home/fuzzing/Desktop/mujs/jsgc.c:17
    #4 0x55678e02b2ad in js_freestate /home/fuzzing/Desktop/mujs/jsgc.c:242
    #5 0x55678e06d43e in main /home/fuzzing/Desktop/mujs/main.c:373
    #6 0x7f7589345b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Direct leak of 2 byte(s) in 2 object(s) allocated from:
    #0 0x7f7589ddaf40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x55678e058126 in js_defaultalloc /home/fuzzing/Desktop/mujs/jsstate.c:12
    #2 0x55678e04a1df in js_free /home/fuzzing/Desktop/mujs/jsrun.c:58
    #3 0x55678e029015 in jsG_freefunction /home/fuzzing/Desktop/mujs/jsgc.c:18
    #4 0x55678e02ad51 in js_gc /home/fuzzing/Desktop/mujs/jsgc.c:187
    #5 0x55678e06c417 in jsB_gc /home/fuzzing/Desktop/mujs/main.c:91
    #6 0x55678e0526f5 in jsR_callcfunction /home/fuzzing/Desktop/mujs/jsrun.c:1041
    #7 0x55678e0530b2 in js_call /home/fuzzing/Desktop/mujs/jsrun.c:1083
    #8 0x55678e056ca2 in jsR_run /home/fuzzing/Desktop/mujs/jsrun.c:1511
    #9 0x55678e052498 in jsR_callscript /home/fuzzing/Desktop/mujs/jsrun.c:1024
    #10 0x55678e052f67 in js_call /home/fuzzing/Desktop/mujs/jsrun.c:1079
    #11 0x55678e058c0d in js_dofile /home/fuzzing/Desktop/mujs/jsstate.c:201
    #12 0x55678e06d304 in main /home/fuzzing/Desktop/mujs/main.c:347
    #13 0x7f7589345b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x7f7589ddaf40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x55678e058126 in js_defaultalloc /home/fuzzing/Desktop/mujs/jsstate.c:12
    #2 0x55678e04a1df in js_free /home/fuzzing/Desktop/mujs/jsrun.c:58
    #3 0x55678e028fd9 in jsG_freefunction /home/fuzzing/Desktop/mujs/jsgc.c:17
    #4 0x55678e02ad51 in js_gc /home/fuzzing/Desktop/mujs/jsgc.c:187
    #5 0x55678e06d432 in main /home/fuzzing/Desktop/mujs/main.c:372
    #6 0x7f7589345b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x7f7589ddaf40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x55678e058126 in js_defaultalloc /home/fuzzing/Desktop/mujs/jsstate.c:12
    #2 0x55678e04a1df in js_free /home/fuzzing/Desktop/mujs/jsrun.c:58
    #3 0x55678e029015 in jsG_freefunction /home/fuzzing/Desktop/mujs/jsgc.c:18
    #4 0x55678e02ad51 in js_gc /home/fuzzing/Desktop/mujs/jsgc.c:187
    #5 0x55678e06d432 in main /home/fuzzing/Desktop/mujs/main.c:372
    #6 0x7f7589345b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x7f7589ddaf40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x55678e058126 in js_defaultalloc /home/fuzzing/Desktop/mujs/jsstate.c:12
    #2 0x55678e04a1df in js_free /home/fuzzing/Desktop/mujs/jsrun.c:58
    #3 0x55678e02908d in jsG_freefunction /home/fuzzing/Desktop/mujs/jsgc.c:20
    #4 0x55678e02b2ad in js_freestate /home/fuzzing/Desktop/mujs/jsgc.c:242
    #5 0x55678e06d43e in main /home/fuzzing/Desktop/mujs/main.c:373
    #6 0x7f7589345b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x7f7589ddaf40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x55678e058126 in js_defaultalloc /home/fuzzing/Desktop/mujs/jsstate.c:12
    #2 0x55678e04a1df in js_free /home/fuzzing/Desktop/mujs/jsrun.c:58
    #3 0x55678e02908d in jsG_freefunction /home/fuzzing/Desktop/mujs/jsgc.c:20
    #4 0x55678e02ad51 in js_gc /home/fuzzing/Desktop/mujs/jsgc.c:187
    #5 0x55678e06c417 in jsB_gc /home/fuzzing/Desktop/mujs/main.c:91
    #6 0x55678e0526f5 in jsR_callcfunction /home/fuzzing/Desktop/mujs/jsrun.c:1041
    #7 0x55678e0530b2 in js_call /home/fuzzing/Desktop/mujs/jsrun.c:1083
    #8 0x55678e056ca2 in jsR_run /home/fuzzing/Desktop/mujs/jsrun.c:1511
    #9 0x55678e052498 in jsR_callscript /home/fuzzing/Desktop/mujs/jsrun.c:1024
    #10 0x55678e052f67 in js_call /home/fuzzing/Desktop/mujs/jsrun.c:1079
    #11 0x55678e058c0d in js_dofile /home/fuzzing/Desktop/mujs/jsstate.c:201
    #12 0x55678e06d304 in main /home/fuzzing/Desktop/mujs/main.c:347
    #13 0x7f7589345b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

SUMMARY: AddressSanitizer: 10 byte(s) leaked in 10 allocation(s).

```

test.js is simple as:

```
var a = [ 1, 2, 3 ]
gc(true)
```

**Comment by ccxvii** [comment]
Why doesn't libfuzzer/asan handle realloc according to the specification?

http://pubs.opengroup.org/onlinepubs/009695399/functions/realloc.html

"If size is 0 and ptr is not a null pointer, the object pointed to is freed."

**Comment by dzonerzy** [comment]
Seems to be a compilation flag https://github.com/llvm-mirror/compiler-rt/blob/master/lib/asan/asan_flags.inc#L154

**Comment by ccxvii** [comment]
The default for that flag is true, so it seems unrelated.

**Comment by sebras** [comment]
@ccxvii your fix, eeea83a807beabb5106b762458c7fdf29970bd60, works for clang, but not for gcc. gcc doesn't have `__has_feature` but instead [defines __SANITIZE_ADDRESS__ to 1](https://stackoverflow.com/questions/34813412/how-to-detect-if-building-with-address-sanitizer-when-building-with-gcc-4-8). How about something along the lines of:

```
--- a/jsstate.c
+++ b/jsstate.c
@@ -7,15 +7,25 @@
 
 #include <assert.h>
 
-static void *js_defaultalloc(void *actx, void *ptr, int size)
-{
+/* Detect ASAN in clang */
 #if defined(__has_feature)
 #if __has_feature(address_sanitizer)
+#define ADDRESS_SANITIZER
+#endif
+#endif
+
+/* Detect ASAN in gcc */
+#if defined(__SANITIZE_ADDRESS__)
+#define ADDRESS_SANITIZER
+#endif
+
+static void *js_defaultalloc(void *actx, void *ptr, int size)
+{
+#if defined(ADDRESS_SANITIZER)
        if (size == 0) {
                free(ptr);
                return NULL;
        }
-#endif
 #endif
        return realloc(ptr, (size_t)size);
 }
```

**Comment by ccxvii** [comment]
Fixed in commit 14dc9355bd71818cf01c1c690c1c91a0978ea9b8.
